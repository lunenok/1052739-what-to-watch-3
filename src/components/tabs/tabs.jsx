import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {getDuration, formatRating, getTextRating, formatDateForReview} from "../../utils.js";
import withActiveTab from "../../hocs/with-active-tab/with-active-tab.jsx";
import {connect} from "react-redux";
import {DataOperation} from "../../reducer/reducer.js";

const TabNames = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`
};

export class Tabs extends PureComponent {

  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {currentMovie} = this.props;
    const id = currentMovie.id;
    this.props.loadReviews(id);
  }

  componentDidUpdate(oldProps) {
    const {currentMovie} = this.props;
    const id = currentMovie.id;
    if (currentMovie.id !== oldProps.currentMovie.id) {
      this.props.loadReviews(id);
    }
  }

  render() {
    const {currentMovie, currentTab, onTabClick, textReviews} = this.props;
    const {genre, rating, director, starring, description, scoresCount, runTime, released} = currentMovie;

    return (
      <div className="movie-card__desc">
        <nav className="movie-nav movie-card__nav">
          <ul className={`movie-nav__list`}>
            {Object.values(TabNames).map((tabName) => {
              return (
                <li key={tabName} className={`movie-nav__item ${currentTab === tabName ? `movie-nav__item--active` : ``}`}>
                  <a
                    href="#"
                    className="movie-nav__link"
                    onClick={(evt) => onTabClick(evt, tabName)}
                  >
                    {tabName}
                  </a>
                </li>
              );
            })}
          </ul>
        </nav>

        {currentTab === TabNames.OVERVIEW && (
          <React.Fragment>
            <div className="movie-rating">
              <div className="movie-rating__score">{formatRating(rating)}</div>
              <p className="movie-rating__meta">
                <span className="movie-rating__level">{getTextRating(rating)}</span>
                <span className="movie-rating__count">{scoresCount} ratings</span>
              </p>
            </div>
            <div className="movie-card__text">
              <p>{description}</p>
              <p className="movie-card__director"><strong>Director: {director}</strong></p>
              <p className="movie-card__starring">
                <strong>
                  Starring:
                  {starring.map((actor) => {
                    return actor;
                  })
                  .join(`, `)}
                </strong>
              </p>
            </div>
          </React.Fragment>
        )}

        {currentTab === TabNames.DETAILS && (
          <React.Fragment>
            <div className="movie-card__text movie-card__row">
              <div className="movie-card__text-col">
                <p className="movie-card__details-item">
                  <strong className="movie-card__details-name">Director</strong>
                  <span className="movie-card__details-value">Wes Andreson</span>
                </p>
                <p className="movie-card__details-item">
                  <strong className="movie-card__details-name">Starring</strong>
                  <span className="movie-card__details-value">
                    {starring.map((actor) => {
                      return actor;
                    }).
                    join(`, \n`)}
                  </span>
                </p>
              </div>
              <div className="movie-card__text-col">
                <p className="movie-card__details-item">
                  <strong className="movie-card__details-name">Run Time</strong>
                  <span className="movie-card__details-value">{getDuration(runTime)}</span>
                </p>
                <p className="movie-card__details-item">
                  <strong className="movie-card__details-name">Genre</strong>
                  <span className="movie-card__details-value">{genre}</span>
                </p>
                <p className="movie-card__details-item">
                  <strong className="movie-card__details-name">Released</strong>
                  <span className="movie-card__details-value">{released}</span>
                </p>
              </div>
            </div>
          </React.Fragment>
        )}

        {currentTab === TabNames.REVIEWS && (
          <React.Fragment>
            <div className="movie-card__reviews movie-card__row">
              <div className="movie-card__reviews-col">

                {textReviews.map((review) => {
                  const {reviewId, reviewUserName, reviewDate, reviewComment, reviewRating} = review;

                  return (
                    <div key={reviewId} className="review">
                      <blockquote className="review__quote">
                        <p className="review__text">
                          {reviewComment}
                        </p>
                        <footer className="review__details">
                          <cite className="review__author">{reviewUserName}</cite>
                          <time className="review__date" dateTime={reviewDate}>
                            {formatDateForReview(reviewDate)}
                          </time>
                        </footer>
                      </blockquote>
                      <div className="review__rating">{formatRating(reviewRating)}</div>
                    </div>
                  );
                })}
              </div>
              <div className="movie-card__reviews-col">
              </div>
            </div>
          </React.Fragment>
        )}
      </div>
    );
  }
}

Tabs.propTypes = {
  currentMovie: PropTypes.shape({
    id: PropTypes.number.isRequired,
    genre: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    scoresCount: PropTypes.number.isRequired,
    director: PropTypes.string.isRequired,
    runTime: PropTypes.number.isRequired,
    released: PropTypes.number.isRequired,
    starring: PropTypes.arrayOf(PropTypes.string).isRequired,
    description: PropTypes.string.isRequired,
    textReviews: PropTypes.arrayOf(PropTypes.shape({
      author: PropTypes.string,
      date: PropTypes.string,
      text: PropTypes.string,
      reviewRating: PropTypes.number
    }))
  }),
  onTabClick: PropTypes.func.isRequired,
  currentTab: PropTypes.string.isRequired,
  loadReviews: PropTypes.func.isRequired,
  textReviews: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    rating: PropTypes.number,
    comment: PropTypes.string,
    date: PropTypes.number,
    user: PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
    }),
  })).isRequired
};

const mapToState = (state) => ({
  textReviews: state.reviews,
});

const mapDispatchToProps = (dispatch) => ({
  loadReviews(id) {
    dispatch(DataOperation.loadReviews(id));
  }
});

export default connect(mapToState, mapDispatchToProps)(withActiveTab(Tabs));
