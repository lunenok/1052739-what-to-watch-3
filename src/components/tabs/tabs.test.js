import React from "react";
import renderer from "react-test-renderer";
import {Tabs} from "./tabs.jsx";
import {Provider} from "react-redux";
import configureStore from 'redux-mock-store';

const mockStore = configureStore([]);

const testMovie = {
  name: `Seven Years in Tibet`,
  posterImage: `https://htmlacademy-react-3.appspot.com/wtw/static/film/poster/Seven_Years_in_Tibet.jpg`,
  previewImage: `https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/seven-years-in-tibet.jpg`,
  backgroundImage: `https://htmlacademy-react-3.appspot.com/wtw/static/film/background/Seven_Years_in_Tibet.jpg`,
  backgroundColor: `#C6CADF`,
  description: `True story of Heinrich Harrer, an Austrian mountain climber who became friends with the Dalai Lama at the time of China's takeover of Tibet.`,
  rating: 3.6,
  scoresCount: 112612,
  director: `Jean-Jacques Annaud`,
  starring: [`Brad Pitt`, `David Thewlis`, `BD Wong`],
  runTime: 136,
  genre: `Adventure`,
  released: 1997,
  id: 1,
  isFavorite: false,
  videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  previewVideoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  textReviews: [
    {
      author: `Kate`,
      date: `2020-04-02T16:42:52.258Z`,
      text: `I didn't find it amusing`,
      reviewRating: 9
    },
    {
      author: `Roman`,
      date: `2020-04-02T16:42:52.258Z`,
      text: `The mannered, madcap proceedings are often delightful, occasionally silly`,
      reviewRating: 3
    },
  ]
};

it(`Should tabs page render correctly`, () => {
  const store = mockStore({
    genre: `All genres`,
    currentMovie: testMovie,
    movieList: []
  });
  const tree = renderer
    .create(
        <Provider store={store}>
          <Tabs
            currentMovie={testMovie}
            loadReviews={()=>{}}
            onTabClick={()=>{}}
            currentTab={`Overview`}
            textReviews={testMovie.textReviews}
          />
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
