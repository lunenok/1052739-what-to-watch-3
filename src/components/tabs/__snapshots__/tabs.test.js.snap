// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should tabs page render correctly 1`] = `
<div
  className="movie-card__desc"
>
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          onClick={[Function]}
        >
          Overview
        </a>
      </li>
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          onClick={[Function]}
        >
          Details
        </a>
      </li>
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          onClick={[Function]}
        >
          Reviews
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="movie-rating"
  >
    <div
      className="movie-rating__score"
    >
      8,9
    </div>
    <p
      className="movie-rating__meta"
    >
      <span
        className="movie-rating__level"
      >
        Very good
      </span>
      <span
        className="movie-rating__count"
      >
         ratings
      </span>
    </p>
  </div>
  <div
    className="movie-card__text"
  >
    <p>
      In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.
    </p>
    <p
      className="movie-card__director"
    >
      <strong>
        Director: 
        Nikita Mikhalkov
      </strong>
    </p>
    <p
      className="movie-card__starring"
    >
      <strong>
        Starring:
        Edward Norton, Jude Law
      </strong>
    </p>
  </div>
</div>
`;
